# 1. Operaciones basicas y tipos de datos
Python es un lenguaje de progranmacion en el que no es necesario declarar el tipo de la variable antes de asignarle un valor como es el caso de C/C++ o Java. Sin embargo, por debajo Python usa diferentes tipos de variables que permiten saber el tipo de dato al que pertenece.

**Nota:** Por definicion cada objeto corresponde a una clase

## 1.1 Tipos de datos y clases internas
Los siguientes son los tipos de datos que son usados de forma nativa por python:

| Tipo | Descripcion |
| ---- | ----------- |
| ``int`` | Un numero entero |
| ``float`` | Un numero de punto flotante |
| ``bool`` | Un balor booleano (True or False) |
| ``str`` | Una cadena de caracteres(texto) |
| ``bytes`` | Una secuencia de bytes |
| ``complex`` | Un numero complejo $a + bj$|
| ``list`` | Una lista de objetos |
| ``tuple`` | Una secuencia ordenada e inmutable de objetos | 
| ``dict`` | Un arreglo que asocia objetos (parejas de Key(llaves) y value(valores)) | 
| ``set`` | Un grupo de objetos sin orden|

Python tambien tienen algunas funciones nativas:

| Funcion | Descripcion |
| ---- | ----------- |
| ``abs(x)`` | El valor absoluto (norma) de un valor complejo o numerico |
| ``sum(x)`` | Suma de todos los valores de un objeto iterable |
| ``max(*args)`` | Valor maximo de una secuencia de valores |
| ``min(*args)`` | Valor minimo de una secuencia de valores |
| ``len(x)`` | Numero de objetos en un arreglo |
| ``type(x)`` | Tipo de dato del objeto |
| ``isinstance(obj, class)`` | Conoce si un objeto ``obj`` es del tipo ``class`` |
| ``bin(x)`` | Convierte un numero en su secuencia binaria |
| ``eval(str)`` | Evalua una instruccion en string y retorna el resultado |
| ``print(*args)`` | Imprime la representacion ``str`` de una secuencia de objetos separados por un espacio |


In [3]:
# Define un entero
a = 1        # Decimal 
b = 0xa      # Hexadecimal
c = 0b01001  # Binario

print(type(a))
print(type(b))
print(type(c))

<class 'int'>
<class 'int'>
<class 'int'>


In [4]:
# Define variables flotantes
a = 1.0
b = 1.33333

print(type(a))
print(type(b))

<class 'float'>
<class 'float'>


In [5]:
# Define variables booleanas
a = True
b = False

print(type(a))
print(type(b))

<class 'bool'>
<class 'bool'>


In [6]:
# Declaracion de variables string
a = "This is a string"
b = 'This is also a string'

print(type(a))
print(type(b))

<class 'str'>
<class 'str'>


In [7]:
# Definicion de numeros complejos
a = 1 + 3j
print(type(a))

<class 'complex'>


In [8]:
# Valor absoluto
print(abs(-3))
print(abs(-3 + 4j))

3
5.0


In [9]:
# Operaciones de maximo y minimo
print(min(-3, 0))
print(max(0, 3))

-3
3


In [10]:
# Obtener la longitud de una cadena de caracteres
print(len('123'))

3


In [11]:
# isinstance
print(isinstance(9 - 4j, complex))

a = False
isinstance(a, float)

True


False

In [12]:
# Imprimir la representacion binaria de un numero
print(bin(255))

0b11111111


In [13]:
# Ejecuta una linea como un codigo de python (Cuidado!)
print(eval('isinstance((3*4) + 7 - 1j, complex)'))

True



## 1.2 Operaciones numericas
Las siguientes operaciones son operaciones que pueden ser realizadas sobre variables numericas, _i.e.,_ ``int``, ``float`` o ``complex``

| Operation | Meaning       |
| --------- | --------------- |
|    +      | Suma           |
|    -      | Resta           |
|    *      | Producto  |
|    /      | Division        |
|    //     | Division con parte entera |
|    \*\*   | Exponencial        |
| ^ | Bitwise XOR |
| &#124; | Bitwise OR |
| & | Bitwise AND |

Por ejemplo las siguientes operaciones pueden ser realizadas:

In [14]:
8/3 # usando decimales

2.6666666666666665

In [15]:
8//3 # unicamente la parte entera

2

In [16]:
2**3 

8

In [17]:
2*3 

6

In [18]:
2 - 3 

-1

In [19]:
2 + 3

5

In [20]:
7 ^ 3

4

In [21]:
4 | 3

7

In [22]:
12 & 3

0

In [23]:
print(bin(1))
print(bin(2))
print(bin(4))
print(bin(16))

0b1
0b10
0b100
0b10000


# Listas

Las listas describen arreglos de datos y son ejemplos de objetos iterables. Estos se definen usando corchetes cuadrados ``[ ]``. Estos pueden contener objetos de diferentes tipos, pudiendo definir una lista como se ve a continuacion:

In [24]:
li = [-5, 7, 4, 9, 1, 12, 2]

li

[-5, 7, 4, 9, 1, 12, 2]

Una de las cosas mas importantes respecto a las listas es la capacidad que tienen, tal como otros objetos iterables, de ser indexados. Es decir, tienen la capacidad de extraer objetos que esten en una posicion determinada usando el indice de dicha posicion (o llave en algunos casos). Es importante recordar que contrario a otros lenguajes en Python los incides comienzan desde 0.

In [25]:
print(li[0])
print(li[3])

-5
9


Si intentamos por ejemplo usar indices que superen el tamaño de la lista, esta nos arrojara un error como se muestra a continuacion:

In [26]:
li[20]

IndexError: list index out of range

Ademas, las listas son modificables, lo que nos permite alterar el contenido de un indice especifico.

In [27]:
li[0] = -100
li

[-100, 7, 4, 9, 1, 12, 2]

Otra operacion que se puede realizar con las listas es lo que se conoce como *slicin* que nos permite hacer la seleccion de una sublista usando la notacion ``lista[a:b]`` es importante notar que el objeto que se encuentre en la posicion del final no sera incluido, a continuacion se pueden ver ejemplos de dicha operacion:

In [28]:
print(li[1:3]) # arreglo nuevo con datos desde el segundo al tercero del arreglo original

print(li[2:6]) # arreglo nuevo con datos desde el tercero al sexto del arreglo original

[7, 4]
[4, 9, 1, 12]


Esta opearcion tambien funciona con un unico indice, ya sea al inicio al final, usando siempre ``:`` como separador. En esos casos se usara desde el indice indicado hasta el final de la lista en el caso que se use ``lista[a:]`` o desde el inicio de la lista hasta el indice indicado en caso que se use con la forma  ``lista[:b]``.

In [29]:
print(li[2:]) # arreglo nuevo con datos desde el tercero valor del arreglo original

print(li[:4]) # arreglo nuevo con datos desde el quinto valor del arreglo original

[4, 9, 1, 12, 2]
[-100, 7, 4, 9]


La indexacion y el *Slicing* tambien funcionan con indices negativos, con los que se puede recorrer el arreglo empezando con el ultimo objeto usando el indice ``-1``: 

In [30]:
print(li)

print(li[-1]) # ultimo dato de la lista

print(li[-3:]) # nuevo arreglo con datos desde el ante-penultimo valor del arreglo original

[-100, 7, 4, 9, 1, 12, 2]
2
[1, 12, 2]


Ademas, existe la posibilidad de invertir una lista usando la siguiente notacion:

In [31]:
print(li)
print(li[::-1])

[-100, 7, 4, 9, 1, 12, 2]
[2, 12, 1, 9, 4, 7, -100]


**Nota:** En terminos generales al usar *slicing* se puede usar la forma ``[a:b:c]``, donde la lista que se retorna es la lista desde ``a`` hasta ``b - 1`` con pasos de ``c``. De este modo, se puede, por ejemplo, tomar unicamente los numeros que esten en indices pares usando la notacion de esta forma:

In [32]:
li[::2]

[-100, 4, 1, 2]

Finalmente, es posible observar que los operadores ``+`` y ``*`` pueden ser usados para concatenar dos listas y replicar una lista, respectivamente. 

In [33]:
print([0, 4, 7] + [2, 3])

print([0,5] * 4)

[0, 4, 7, 2, 3]
[0, 5, 0, 5, 0, 5, 0, 5]


## Orden

Para ordenar listas y objetos iterables, python nos ofrece dos funciones nativas que pueden ser usadas. La diferencia entre ambos depende mas en cuanto a terminos del objeto que se retorna al finalzar dicha funcion.

La funcion ``sorted()`` retorna un **nuevo** objeto con la informacion ordenada de la lista que se ingresa por parametro. Por su lado ``lista.sort()`` modifica la lista sobre la cual se corre dicha funcion, de tal forma que la lista con el orden original ya no puede ser usada. por ejemplo:

In [34]:
example_list = [-1, 0, 100, -29, 2, 3, 6]

Primero, obtendremos una nueva lista usando ``sorted``:

In [35]:
ordered_list = sorted(example_list)
print(ordered_list)
print(example_list,'como pueden ver no altero la lista original') 

[-29, -1, 0, 2, 3, 6, 100]
[-1, 0, 100, -29, 2, 3, 6] como pueden ver no altero la lista original


Ahora, ordenaremos al lista modificando la lista original:

In [36]:
example_list.sort()
print(example_list, 'en este caso si fue alterada')

[-29, -1, 0, 2, 3, 6, 100] en este caso si fue alterada


## Problemas

### Problema 1

Construir una lista que tenga `100` repeticiones de los numeros `-1`, `0` y `1`, en ese orden exacto.

In [162]:

lista_1 = '' # Lista que contendra la solucion (no cambien el nombre)
# YOUR CODE HERE
lista_1 =[-1, 0, 1]*100

In [163]:
print(lista_1)

[-1, 0, 1, -1, 0, 1, -1, 0, 1, -1, 0, 1, -1, 0, 1, -1, 0, 1, -1, 0, 1, -1, 0, 1, -1, 0, 1, -1, 0, 1, -1, 0, 1, -1, 0, 1, -1, 0, 1, -1, 0, 1, -1, 0, 1, -1, 0, 1, -1, 0, 1, -1, 0, 1, -1, 0, 1, -1, 0, 1, -1, 0, 1, -1, 0, 1, -1, 0, 1, -1, 0, 1, -1, 0, 1, -1, 0, 1, -1, 0, 1, -1, 0, 1, -1, 0, 1, -1, 0, 1, -1, 0, 1, -1, 0, 1, -1, 0, 1, -1, 0, 1, -1, 0, 1, -1, 0, 1, -1, 0, 1, -1, 0, 1, -1, 0, 1, -1, 0, 1, -1, 0, 1, -1, 0, 1, -1, 0, 1, -1, 0, 1, -1, 0, 1, -1, 0, 1, -1, 0, 1, -1, 0, 1, -1, 0, 1, -1, 0, 1, -1, 0, 1, -1, 0, 1, -1, 0, 1, -1, 0, 1, -1, 0, 1, -1, 0, 1, -1, 0, 1, -1, 0, 1, -1, 0, 1, -1, 0, 1, -1, 0, 1, -1, 0, 1, -1, 0, 1, -1, 0, 1, -1, 0, 1, -1, 0, 1, -1, 0, 1, -1, 0, 1, -1, 0, 1, -1, 0, 1, -1, 0, 1, -1, 0, 1, -1, 0, 1, -1, 0, 1, -1, 0, 1, -1, 0, 1, -1, 0, 1, -1, 0, 1, -1, 0, 1, -1, 0, 1, -1, 0, 1, -1, 0, 1, -1, 0, 1, -1, 0, 1, -1, 0, 1, -1, 0, 1, -1, 0, 1, -1, 0, 1, -1, 0, 1, -1, 0, 1, -1, 0, 1, -1, 0, 1, -1, 0, 1, -1, 0, 1, -1, 0, 1, -1, 0, 1, -1, 0, 1, -1, 0, 1, -1, 0, 1, -1, 0, 1]

### Problema 2

Construir una lista que tenga un `1` rodeado de `20` ceros a cada lado.

In [40]:
lista_2 = '' # Lista que contendra la solucion (no cambien el nombre)
# YOUR CODE HERE
lista_2 = [0]*20+[1]+[0]*20

In [41]:
print(lista_2)

[0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0]


### Problema 3

Calcular el promedio de la siguiente lista:

In [48]:
li = [3, 18, 17, 44, 14, 12, 29, 19, 4, 6, 17, 7, 14, 6, 8, 17, 17, 21, 65,\
      19, 10, 31, 92, 17, 5, 15, 3, 14, 20, 12, 29, 57, 15, 2, 17, 1, 6, 17, 2,\
      71, 12, 11, 62, 14, 9, 20, 43, 19, 4, 15]

prom = '' # variable que debe contener el promedio(no cambiarle el nombre)
# YOUR CODE HERE
prom = sum(li)/len(li)

In [49]:
print(f'promedio: {prom}')

promedio: 20.04


*Respuesta*:

    20.04

# Tuplas

Las tuplas son listas de objetos con la caracteristica de ser inmutables y ordenadas. Mantienen las mismas operaciones de ``slicing`` e indexacion, sin embargo, no se pueden modificar los valores. Definimos las listas usando parentesis en lugar de corchetes cuadrados:

In [50]:
tp = (1, 2, 3, 4, 'a')

print(tp[3])

print(tp[-1])

print(tp[2:])

4
a
(3, 4, 'a')


Recuerden que las tuplas son inmutables:

In [51]:
tp[2] = 'b'

TypeError: 'tuple' object does not support item assignment

**Nota**: Podemos definir tuplas sin la necesidad de utilizar el parentesis, tal como se ve en la siguiente celda

In [52]:
tp1 = 'a', 'b', 2

tp1

('a', 'b', 2)

# Desempaquetamiento

En python podemos "desempaquetar" los objetos contenidos en un objeto iterable (lista, tupla, diccionario, etc.) usando la notacion ``var1, var2, ..., varN = iterable``. Para lograrlo es necesario que el objeto iterable contenga ``N`` elementos, de lo contrario arrojara una excepcion:

In [53]:
# desempaquetar listas
a, b, c = [2, 3, 4] 
print(a, b, c)

# desempaquetar strings
a, b, c = '123'
print(a, b, c)

# desempaquetar tuplas
a, b, c = 4, -2, 1+9j
print(a, b, c)

2 3 4
1 2 3
4 -2 (1+9j)


In [54]:
# menos elementos que variables
a, b, c = [2, 3]

ValueError: not enough values to unpack (expected 3, got 2)

In [55]:
# Menos variables que elementos
a, b = [4, 5, 6] 

ValueError: too many values to unpack (expected 2)

In [56]:
# algunos trucos que son posibles!
(a, b), c = [(3, 4), 'jasdlkasd']
print(a, b, c)

3 4 jasdlkasd


## Problemas

### Problema 1

Es posible calcular el promedio a la lista de la siguiente tupla?

In [58]:
li = (3, 18, 17, 44, 14, 12, 29, 19, 4, 6, 17, 7, 14, 6, 8, 17, 17, 21, 65,\
      19, 10, 31, 92, 17, 5, 15, 3, 14, 20, 12, 29, 57, 15, 2, 17, 1, 6, 17, 2,\
      71, 12, 11, 62, 14, 9, 20, 43, 19, 4, 15)

prom = '' # variable que debe contener el promedio(no cambiarle el nombre)
prom = sum(li)/len(li)

In [59]:
print(f'promedio: {prom}')

promedio: 20.04


### Problema 2

Crear una tupla que tenga un solo elemento

In [63]:
tup = '' # variable que debe contener la tupla(no cambiarle el nombre)
# YOUR CODE HERE
tup = (1,)

In [64]:
print(f'Tipo: {type(tup)}')

Tipo: <class 'tuple'>


# Diccionarios

Los diccionarios son uno de los tipos mas usados de datos en Python. Hasta el momento hemos visto tipos de datos iterables que estan indexados con numeros enteros. Los diccionarios, contrario a las listas, pueden usar cualquier tipo de dato como indice el cual resulta ser **inmutable**. Esto permite realizar un arreglo asociativo donde el objeto se obtiene usando la llave que le corresponde. 

Por ejemplo, esto permitiria usar una lista que tenga contraseñas asociadas a usuarios. Ya que tenemos la asociacion del usuario como llave y la contraseña como valor, podremos obtener la contraseña de cada usuario al usar su nombre como llave.

In [72]:
codes = {'Luis': 2257, 'Juan': 9739, 'Carlos': 5591}

Como pueden ver, se delcara un diccionario usando corchetes cuadrados y dos puntos dividiendo los valores de las llaves ``{k:v}`` 

Para recuperar un valor usando su llave como si fuera el indice de una lista:

In [66]:
codes['Carlos']

5591

In [67]:
codes['Juan']

9739

El diccionario tambien permite realizar cambios a los valores asociados, como si alguien cambiara su contraseña:

In [74]:
codes['Luis'] = 1627

In [75]:
codes

{'Luis': 1627, 'Juan': 9739, 'Carlos': 5591}

**Nota**: Los diccionarios no son ordenados, lo que quiere decir que como tal no existe un orden entre las llaves al momento de la definicion de estos tipos de datos. Para sobrellevar esta limitacion existe la clase ``OrderedDict`` dentro del paquete ``collections`` que ofrece las mismas operaciones que un diccionario ordinario.

Otra operacion que se puede realizar es eliminar elementos de un diccionario, el cual dado el caso que estamos llevando, seria equivalente a que alguien cancele la subscripcion. Los comandos para esto son ``pop`` o ``del`` que se usan de la siguiente forma:

In [76]:
val = codes.pop('Juan')
print(val)

del codes['Luis']
print(codes)

9739
{'Carlos': 5591}


Si quisieramos agregar una nueva llave al diccionario podemos usar la misma notacion que usamos para llamar las variables pero en este caso asignandole un valor.

In [77]:
codes['Jorge'] = 6621

codes

{'Carlos': 5591, 'Jorge': 6621}

Finalmente para lograr obtener todas las llaves o valores del diccionario podemos usar las funciones internas ``keys`` y ``values`` respectivamente:

In [78]:
codes.keys()

dict_keys(['Carlos', 'Jorge'])

In [79]:
codes.values()

dict_values([5591, 6621])

## Problemas

Dado el siguiente diccionario que guarda las notas de distintos estudiantes

In [80]:
notas = {
    'Juan': [4.5, 3.7, 3.4, 5],
    'Alicia': [3.5, 3.1, 4.2, 3.9],
    'German': [2.6, 3.0, 3.9, 4.1]
}

### Problema 1
calcular:

* La nota promedio de Juan (recuerde que se puede utilizar `sum` y `len` para obtener el promedio).

*Respuesta*

    4.15

In [90]:
prom_juan = '' # variable que debe contener el promedio de Juan
# YOUR CODE HERE
prom_juan = sum(notas['Juan'])/len(notas['Juan'])

In [91]:
print(f'el promedio de Juan es: {prom_juan}')

el promedio de Juan es: 4.15


* La nota promedio del curso

*Respuesta*

    3.74

In [101]:
prom_class = '' # variable que debe contener el promedio general
# YOUR CODE HERE
glob = []
for i in notas.values():
    glob += i
prom_class = sum(glob)/len(glob)


In [102]:
print(f'promedio de la clase: {prom_class}')

promedio de la clase: 3.7416666666666667


# Conversion de tipos

Para poder hacer la conversion de un tipo de dato a otro podemos usar las *keywords* de los tipos de datos como funcion:

* **int**: Conversion a un tipo de dato int(numeros enteros)

In [103]:
print(int(3.99))

print(int('6'))

3
6


* **float**: Conversion a punto flotante

In [104]:
print(float(12))

print(float('4.23'))

12.0
4.23


* **str**: La representacion en string del tipo de objeto (puede ser variado y personalizable al hacer una clase)

In [105]:
print(str(36.1))

print(str([1,2,3]))

36.1
[1, 2, 3]


* **list**: Convertir objetos iterables en una lista

In [106]:
print(list((3, 2, 4)))

print(list('1457'))

[3, 2, 4]
['1', '4', '5', '7']


Al hacer esto con un diccionario solo mantendra las llaves.

In [107]:
list({'a': 12, 'b': 5})

['a', 'b']

* **dict**: Podemos realizarlo al tener parejas de llaves y valores emparejados en una lista (K, V).

In [108]:
dict([[10, 'a'], [15, 't']])

{10: 'a', 15: 't'}

Tambien podemos usar ``zip`` para crear un diccionario basado en una lista de llaves y una lista de valores:

In [109]:
keys = ['A', 'B', 'C']
values = [12, 34, -1 + 9j]
pairs = list(zip(keys, values))
print(pairs)

dict(pairs)

[('A', 12), ('B', 34), ('C', (-1+9j))]


{'A': 12, 'B': 34, 'C': (-1+9j)}

# Funciones

Como en cualquier otro lenguaje, en *Python* tambien es posible definir funciones, es decir, secuencias de enunciados que reciben ciertos datos, ejecutan algunas operaciones sobre ellos y devuelven un resultado.

Para definir una funcion se usa la palabra clave `def`, y el valor que va a retornar siempre debe ser precedido por un `return`. La sintaxis de una funcion es como se ve a continuacion:

```python
def NOMBRE(LISTA DE ARGUMENTOS):
    ENUNCIADOS
    return VALOR
```

o

```python
def NOMBRE(LISTA DE ARGUMENTOS):
    ENUNCIADOS
    print(VALOR)
```

La linea que contiene el `return` (o `print`) es opcional, pues no todas las funciones deben retornar algo. Por ejemplo, hay algunas que solo modifican los valores de ciertas variables, por lo que no necesitan retornar o imprimir ningun valor.

**Nota**:

Es muy importante tener en cuenta que los enunciados que hacen parte de la funcion deben estar **una indentacion** por dentro del encabezado. En otras palabras, todo lo que este indentado por dentro de la definicion, pertenece al cuerpo de la funcion, ya que en Python la indentacion es lo unico que define la forma en que se agrupa el codigo. Solo cuando el nivel de indentacion se retorne al punto en que se escribio el primer `def` se considera que ha terminado la definicion de la funcion.

Un ejemplo muy sencillo de una funcion que toma un argumento `x` y retorna este argumento elevado al cuadrado es:

In [110]:
def cuadrado(x):
    return x**2

Podemos comprobar que la funcion esta operando correctamente al pasarle varios argumentos y ver los resultados que retorna:

In [111]:
cuadrado(3)

9

In [112]:
cuadrado(5)

25

In [113]:
cuadrado(10)

100

In [114]:
cuadrado('a')

TypeError: unsupported operand type(s) for ** or pow(): 'str' and 'int'

En el ultimo caso vemos que si intentamos pasarle a la funcion un argumento que no puede ser procesado, Python simplemente retorna un error.

## Parametros opcionales

En ocasiones, se desea que una funcion cuente con unos parametros por defecto. Por ejemplo si se desean imprimir mensajes de ejecucion en una funcion, u otro valor numerico que cuenta con una inicializacion especifica para una rutina de calculo, estos parametros se definen empleando la sintaxis ``arg=value``, donde ``arg`` es el nombre del argumento y ``value``, el valor de inicializacion por defecto. 

A continuacion se presenta un ejemplo de una funcion que recibe argumentos opcionales.

In [115]:
def suma(x, y, verbose=False):
    if verbose:
        print("Sumara {0} y {1}".format(x, y))
    return x + y

Como es posible apreciar, la funcion recibe un parametro opcional ``verbose``, el cual permite imprimir los argumentos a sumar, ahora procedemos a invocar esta funcion, primero sin el parametro opcional:

In [116]:
suma(-1, 2.5)

1.5

Ahora, podemos cambiar el valor del parametro opcional con el fin de observar los argumentos a sumar:

In [117]:
suma(-1, 2.5, verbose=True)

Sumara -1 y 2.5


1.5

Es posible definir multiples argumentos opcionales, como por ejemplo en la siguiente funcion que imprime el nombre de una persona:

In [118]:
def imprimir_nombre(primer_nombre, segundo_nombre='', apellidos=''):
    print("Nombre: ", primer_nombre, segundo_nombre, apellidos)

Ahora veamos los distintos resultados de esta funcion

In [119]:
imprimir_nombre('Carlos')

Nombre:  Carlos  


In [120]:
imprimir_nombre('Carlos', segundo_nombre='Andres')

Nombre:  Carlos Andres 


In [121]:
imprimir_nombre('Carlos', segundo_nombre='Andres', apellidos='Cordoba Chaves')

Nombre:  Carlos Andres Cordoba Chaves


Tambien podemos definir funciones recursivas:

In [122]:
def fib(N, F0=0, F1=1):
    if N == 0:
        return F1
    return fib(N - 1, F1, F0 + F1)

print(fib(40))

165580141


## Funciones internas

Python permite definir funciones al interior de otra funcion, las cuales se denominan funciones internas. Estas funciones pueden acceder al espacio de variables de la funcion que la encapsula. Sin embargo, estas funciones no son accesibles desde el exterior de la funcion.

Por ejemplo, a continuacion se presenta una funcion que toma por parametro un elemento y retorna una lista con distintas operaciones sobre el elemento inicial:

In [123]:
def doblar_y_dividir_por_dos(x):
    def doblar():
        return 2 * x

    def dividir_por_dos():
        return x / 2

    return [doblar(), dividir_por_dos()]

In [124]:
doblar_y_dividir_por_dos(10)

[20, 5.0]

Si se intenta invocar a las funciones internas directamente, veremos que aparece un error debido a estas no existen por fuera de `doblar_y_dividir_por_dos`:

In [125]:
doblar

NameError: name 'doblar' is not defined

In [126]:
dividir_por_dos

NameError: name 'dividir_por_dos' is not defined

## Funciones anonimas

Inspirado en los lenguajes funcionales, Python tambien permite definir funciones anonimas, las cuales corresponden a funciones que se pueden asignar a variables o pueden ser empleadas en los argumentos de otras funciones que requieran una funcion por parametro. Estas funciones se crean empleando la palabra clave `lambda`, como es posible apreciar a continuacion:

In [127]:
f = lambda x, y: x + y

f(2, 3)

5

En el contexto de otras funciones, es posible recordar la funcion `sorted`, introducida en el modulo de listas. Esta funcion ademas de ordenar elementos sencillos, tambien puede ordenar listas que contienen elementos mas complejos (por ejemplo diccionarios) empleando un criterio especifico tal como el valor de una llave especifica a traves del argumento `key`. Este argumento espera una funcion que tiene como parametro un elemento de la lista, y retorna el valor por el cual debe ordenarse el elemento en la lista:

In [128]:
lista_desordenada = [{'nombre': 'enlatado', 'precio': 1000},
                     {'name': 'jamon', 'precio': 2000},
                     {'nombre': 'huevos', 'precio': 200}]

lista_ordenada_por_precio = sorted(lista_desordenada, key=lambda x: x['precio'])

lista_ordenada_por_precio

[{'nombre': 'huevos', 'precio': 200},
 {'nombre': 'enlatado', 'precio': 1000},
 {'name': 'jamon', 'precio': 2000}]

**Nota:** Las funciones anonimas solo pueden ser empleadas para realizar operaciones basicas, asi como invocacion de funciones. Sin embargo, no es posible realizar flujos de control mas complejos (e.g. ciclos `for`, `while`, etc).

## Problemas

### Problema 1

Definir una funcion `imprimir_doble` que tome un argumento `x` y lo imprima dos veces, con un espacio entre una palabra y la siguiente. Por ejemplo, al evaluarla debe retornar:

```python
imprimir_doble(5)
5 5
```
     
```python
imprimir_doble('hola')
hola hola
```
     
```python
imprimir_doble([3,9,4])
[3,9,4] [3,9,4]
```

In [129]:
# Escribir la solucion aqui
def imprimir_doble(x):
    print(x, x)

In [130]:
imprimir_doble(5)
imprimir_doble('hola')
imprimir_doble([3,9,4])

5 5
hola hola
[3, 9, 4] [3, 9, 4]


### Problema 2

Definir una funcion `distancia` que tome dos argumentos `x,y`, que sean listas de dos elementos, y calcule la distancia entre ellos usando el teorema de Pitagoras:

$$\sqrt{\left(x_{1}-y_{1}\right)^{2}+\left(x_{2}-y_{2}\right)^{2}}$$

Pueden comprobar que la funcion esta haciendo su trabajo correctamente si retorna estos valores:

```python
distancia([0,0], [1,1])
1.4142135623730951
```

```python
distancia([1,5], [2,2])
3.1622776601683795
```

In [131]:
def distancia(p1,p2):
    # YOUR CODE HERE
    return ((p1[0]-p2[0])**2 + (p1[1]-p2[1])**2)**(1/2)

In [132]:

print(distancia([0,0], [1,1]))
print(distancia([1,5], [2,2]))



1.4142135623730951
3.1622776601683795


### Problema 3

Definir una funcion `digitos` que tome un numero `x` y retorne los digitos de que se compone, como cadenas. Por ejemplo, `digitos` debe retornar:

```python
digitos(1234)
['1', '2', '3', '4']
```

```python
digitos(99861)
['9', '9', '8', '6', '1']
```

**Nota**: Utilizar los comandos de conversion entre tipos de datos, vistos al final de la seccion anterior

In [133]:
def digitos(x):
    # YOUR CODE HERE
    return list(str(x))

In [134]:
# Evaluar la funcion aqui

digitos(1234)
digitos(99861)



['9', '9', '8', '6', '1']

# Cadenas de caracteres

Los strings corresponden a una lista de caracteres, por lo tanto son iterables, lo que nos permite aplicar slicin e indexacion como lo aprendimos anteriormente. Como se vio con anterioridad, los strings se pueden definir con una comilla simple o doble:

In [135]:
fruit = "banana"

In [136]:
dulce = 'bocadillo'

With respect to common operators such as * and +, we can obtain the following results:
En este caso, los operadores + y * dan los siguientes resultados:

| Operador | Dato de input           | Resultado                                         
| --------- | --------------- | --------- 
|    +      | str + str | Concatena ambos str                                  
|    \*     | str * int | Concatena el str la cantidad int idicada 

Podemos usar estos operadores con nuestros datos:

In [137]:
fruit + dulce

'bananabocadillo'

In [138]:
fruit * 3

'bananabananabanana'

In [139]:
dulce[0]

'b'

In [140]:
dulce[:7]

'bocadil'

Sin embargo, a diferencia de las listas, los strings son inmutables, lo que significa que no podemos asignar valores nuevos a la cadena de caracteres:

In [141]:
fruit[2] = 'z'

TypeError: 'str' object does not support item assignment

Los strings tienen muchos metodos asociados. Podemos ver varios de estos al agregar un punto <kbd>('.')</kbd> al string y presionar <kbd>Tab</kbd>. Por ejemplo, veamos las posibles operaciones sobre nuestra variable `fruit`:

In [145]:
fruit.capitalize()

'Banana'

**Nota**: Todos los metodos retornan un nuevo string, puesto que estos son inmutables.
Aqui veremos algunos de los mas importantes metodos para la manipulacion de strings:

* **upper**: Convierte todos los caracteres en mayusculas

In [146]:
fruit.upper()

'BANANA'

* **count**: Cuenta el numero de apariciones de un caracter en un string

In [147]:
fruit.count('a')

3

* **replace**: Reemplaza un caracter por otro dentro del string

In [148]:
fruit.replace('a', 'o')

'bonono'

In [149]:
fruit.replace('ban', 'en')

'enana'


* **split**: Separa el string por espacios y retorna una lista que contiene cada palabra individual.

In [150]:
s = "Hello, world! hello world"

In [151]:
s.split()

['Hello,', 'world!', 'hello', 'world']


Tambien podemos dividir las palabras usando otro caracter de string:

In [152]:
dulce.split('ca')

['bo', 'dillo']

* **rstrip:** Eliminar cualquier carácter de espacio nuevo final al final de una cadena

In [156]:
s = 'I want my line without any newline\n'
s.rstrip()

'I want my line without any newline'


Tambien funciona con el fin de linea de Windows (CRLF):

In [157]:
s = 'Windows is the original sin  \r\n'
s.rstrip()

'Windows is the original sin'

# Strings formateados

In [158]:
duenio = 'Laura'
mascota = 'Max'
edad = 14.576
mi_string = f'La mascota de {duenio} se llama {mascota} y tiene {edad:3.2f} anios'
print(mi_string)

La mascota de Laura se llama Max y tiene 14.58 anios


In [159]:
fisico = 'Boltzmann'
constante = 1.38064 * 10**(-23)
# o, mas comodamente
constante = 1.38064e-23
mi_string = f'La constante de {fisico} tiene un valor de {constante:3.2E}'
print(mi_string)

La constante de Boltzmann tiene un valor de 1.38E-23


## Problemas

### Problema 1

Tomar la variable `dulce`, hacer que se repita 50 veces, y separar las palabras con un espacio, de tal forma que obtengamos algo como lo siguiente, pero **sin** generar un espacio al final.

        'bocadillo bocadillo ...'

In [ ]:
# Escribir la solucion aqui


### Problema 2

Cuuntas veces se repite la palabra `banano` en la siguiente cadena?:

In [ ]:
muchas_frutas = 'banAnobanAnobananobanaNobananobananobanaNobaNanobanano\
bananobananobaNanobananobananobaNanobAnanobananobananobanaNobananobanAno\
bananobananobanaNobananobananobananobananobananobananobananobananobAnAno\
bAnanobananobananobananobananobananobanANobananobananobanaNobananobanano\
bananobanaNobAnAnobananobananobananobananobananobAnAnobananobananobanano\
baNanobananobananobaNaNobananobANanobananobananobananobAnanobananobanano\
bananobananobAnanobananobaNAnobananobananobananobaNanobanaNobANanobanano\
baNanobananobananobAnanobananobananobananobaNAnobananobanANobananobAnano\
bANanobanAnobananobaNanobananobananobananobananobananobananobAnanobanano\
bananobanAnobananobananobanAnobananobananobananobanAnobananobananobaNano\
bAnanobananobAnanobaNanobananobanaNobananobananobanANobananobananobANAno\
bananobananobaNAnobanaNobAnanobanAnobananobananobanAnobaNanobananobanaNo\
banaNobANAnobananobananobanAnobananobananobanANobananobanAnobananobanano\
banaNobananobAnanobananobAnanobananobanANobananobananobanAnobanaNobanano\
bananobAnanobananobaNanobananobanANobananobananobananobaNAnobananobanAno\
bananobananobananobaNanobananobananobanAnobananobananobANanobananobanano\
bananobananobaNanobananobananobananobAnanobananobananobananobananobanano\
bananobanANobananobanaNobAnanobananobaNanobaNAnobananobananobananobanano\
bananobananobananobananobananobAnanobanaNobananobananobaNAnobananobanANo\
bananobanaNobananobananobananobananobananobaNanobananobanaNobanAnobanAno\
bananobanAno'

*Respuesta*:

    150

In [ ]:
# Escribir la solucion aqui


### Problema 3

Cuantas veces se repite `banano` en la cadena anterior, sin importar si algunas de sus letras estan en mayusculas o no?

*Respuesta*:

    239

In [ ]:
# Escribir la solucion aqui


# Importar librerias

Python posee un gran variedad de librerias para todo tipo de usos, desde la creacion y apertura de archivos en cualquier sistema operativo hasta la extraccion de datos en archivos de Word o Excel, pasando por el manejo y analisis de enormes cantidades de informacion provenientes de la bolsa de valores o de experimentos astronomicos, climaticos o biologicos.

La ventaja de Python sobre otros lenguajes de programacion es que posee un avanzado mecanismo para importar librerias o comandos especificos de alguna de ellas, lo que hace muy sencillo usarlas para beneficio del programador. A continuacion veremos las distintas formas en que puede hacerse esto.

## Usando `import`

La primera opcion, y la mas sencilla, es usar el comando `import` seguido del nombre la libreria, asi:

```python
import math
```

Cuando querramos usar algun comando de la libreria, debemos llamarla por su nombre seguido de un punto (`.`) y la tecla <kbd>Tab</kbd>. Inmediatamente deberi aparecer una lista con las funciones que hacen parte de la misma, asi:

In [ ]:
import math

Como `math` es la libreria de funciones matematicas basicas de Python, podemos utilizarla para calcular el $\textrm{sin}(\pi/2)$, por ejemplo:

In [ ]:
math.sin(math.pi/2)

## Usando `from`

Otra opcion es solo llamar un comando especifico de una libreria determinada. Para ello usamos la siguiente sintaxis:

```python
from math import sin
```

En este caso solo se carga el comando `sin`, sin cargar el resto de la libreria. Esto puede ser muy ventajoso cuando la libreria es grande y solo queremos utilizar una pequeña funcionalidad de la misma. Otra ventaja es que no es necesario usar el nombre de la libreria antes del comando.

In [ ]:
from math import sin

In [ ]:
sin(3.1)

Sin embargo, si queremos utilizar `pi`, veremos un error en la consola, porque no ha sido cargado:

In [ ]:
sin(pi/2)

Para no tener que escribir una linea `from math import ...` por cada simbolo que se desee importar, se puede usar la misma sintaxis de arriba con cada simbolo separado por una coma, asi:

In [ ]:
from math import sin, pi

Al cargar `pi` de `math` de esta forma, ya no veremos el error anterior:

In [ ]:
sin(pi/2)

## Usando un nombre corto

Una opcion para no tener que llamar una libreria por su nombre completo, es renombrarla al momento de importarla. Esto es una practica bastante comun, pues permite utilizar nombres cortos para las librerias, lo que hace mucho mas facil llamarlas. Para ello se usa la siguiente sintaxis:

```python
import math as mt
```

De esta forma, en lugar de `math` solo usamos `mt`, asi:

In [ ]:
import math as mt

In [ ]:
mt.sin(mt.pi/2)

## Importando todos los comandos

La ultima opcion consiste en llamar todos los comandos de una libreria sin que esten precedidos por su nombre. Para ello se utiliza la sintaxis:

```python
from math import *
```

**Advertencia**:

Esta opcion no es muy recomendable ya que los comandos de varias librerias pueden tener el mismo nombre, lo que puede resultar en serias confusiones.

Un ejemplo de este caso es el siguiente:

In [ ]:
from math import *

In [ ]:
cos(pi/2)

## Problemas

### Problema 1

Importar la libreria `numpy` con el nombre `np`. Despues utilizar la funcion `arange` de la misma para generar una lista de numeros entre `0` y `1`, en pasos de `0.01`.

**Nota**:

Para mirar la ayuda de `arange`, se debe evaluar una celda nueva con la siguiente sintaxis:

```python
np.arange?
```

In [ ]:
# Escribir la solucion aqui


### Problema 2

De la misma libreria anterior, importar unicamente la funcion `linspace` y la constante `pi`. Utilizar estas dos expresiones para generar una lista de 100 numeros entre $0$ y $2 \pi$.

In [ ]:
# Escribir la soluci�n aqu�


### Problema 3

Usar la funcion `sin` de `numpy` para calcular el seno de los numeros generados por `linspace` en el ejercicio anterior.

**Notas**:

1. **No** es necesario usar un ciclo `for`. Mirar detenidamente los ejemplos de `numpy.sin` en el visor de ayuda.

2. Puede hacerse lo mismo con `math.sin`?

In [ ]:
# Escribir la solucion aqui


### Problema 4

Generar una lista de 50 puntos aleatorios en 2D, usando la funcion `rand` de la libreria `numpy.random`

In [ ]:
# Escribir la solucion aqui


# If condicional

Un clasico de todos los lenguajes de programacion completos de Turing. En Python se define usando la siguiente sintaxis: 

In [ ]:
def lesser_or_greater(x, y):
    if x < y:
        print(x, "is lesser than", y)
    elif x > y:
        print(x, "is greater than", y)
    else:
        print(x, "and", y, "are equal")


Si invocamos la funcion usando diferentes inputs, obtendremos diferentes outputs:

In [ ]:
lesser_or_greater(2, 5)

In [ ]:
lesser_or_greater(100, 10)

In [ ]:
lesser_or_greater(1, 1)

Algunos operadores de comparacion importantes estan enlistados en la siguiente tabla:

| Operador | Afirmacion
| -------- | ---------
|   ==     | Igualdad
|   !=     | Desigualdad        
|   <      | Menor que
|   >      | Mayor que
|   <=     | Menor o igual
|   >=     | Mayor o igual
|   not    | Negativo de una condicion booleana
|   and    | Operacion AND entre dos booleanos
|   or     | Operacion OR entre dos booleanos
|   in     | Verifica si un elemento esta dentro de un objeto iterable
|   is     | Verifica si un objeto es igual a otro (basado en memoria)

Todos los operadores son similares a otros lenguajes de programacion, con excepcion de los operadores ``in``, ``is`` and ``not``:

In [ ]:
3 in [1, 2, 4]

In [ ]:
3 in [1, 2, 3]

In [ ]:
'a' in {'a':4, 'b':3}

In [ ]:
3 in (25, 6)

In [ ]:
not (2 == 5)

In [ ]:
3 is None

In [ ]:
4 is not None

A traves de estos ejemplos tambien podemos notar que los valores de verdad en Python se escriben como `True` y `False` para verdadero y falso, respectivamente.

# Asignacion condicional

In [ ]:
condicion = False
if condicion:
    variable = 'si!'
else:
    variable = 'no!'
print(variable)

In [ ]:
variable = 'si!' if condicion else 'no!'
print(variable)

## Problemas

### Problema 1

Definir una funcion `absoluto(x)` que tome un numero `x` y retorne su valor absoluto, asi:

```python
absoluto(100.22)
100.22
```

```python
absoluto(-18.7)
18.7
```

In [ ]:
# Escribir la solucion aqui
def absoluto(x):
    # YOUR CODE HERE
    raise NotImplementedError()

In [ ]:
assert(absoluto(100.22)==100.22)
assert(absoluto(-18.7)==18.7)

### Problema 2

Definir una funcion `es_divisible_entre_siete(x)` que imprima si un numero es o no es divisible entre 7. La funcion debe *imprimir* resultados como los siguientes:

```python
es_divisible_entre_siete(12)
'12 no es divisible entre 7'
```

```python
es_divisible_entre_siete(14)
'14 es divisible entre 7'
```

```python
es_divisible_entre_siete(32)
'32 no es divisible entre 7'
```

```python
es_divisible_entre_siete(21)
'21 es divisible entre 7'
```

**Nota**:

Utilizar el operador modulo (`%`) para decidir si un numero es moltiplo de otro. Este operador retorna el resto de la division entre dos numeros. Por tanto, si un numero divide exactamente a otro, retorna `0`, sino retorna cualquier otro numero. Veamos algunos ejemplos:

```python
12%4
0
```

```python
12%6
0
```

```python
12%5
2
```

```python
25%5
0
```

```python
25%6
1
```

In [ ]:
# Escribir la solucion aqui


### Problema 3

Generalizar la funcion anterior como una nueva funcion llamada `es_divisible_entre_n(x, n)` que tome dos n�meros enteros e imprima si el primero es divisible entre el segundo, asi: (Tomado de *Aprenda a pensar como un programador con Python*)

```python
es_divisible_entre_n(20, 4)
'20 es divisible entre 4'
```

```python
es_divisible_entre_n(36, 5)
'36 no es divisible entre 5'
```

In [ ]:
# Escribir la solucion aqui


### Problema 4

Definir una funcion `agregar_nuevo(li, x)` que reciba una lista y un elemento, y retorne una nueva lista en la que esta añadido el elemento, pero solo si este **no** hace parte de la lista original (Tomado de *Introduccion a Mathematica* del Prof. Jurgen Tischer).

Por ejemplo, algunos resultados de esta funcion son:

```python
agregar_nuevo([3,9,6], 11)
[3, 9, 6, 11]
```

```python
agregar_nuevo([3,9,6], 9)
[3, 9, 6]
```

In [ ]:
# Escribir la solucion aqui


# Ciclo For

En Python, el ``for .. in`` es usado para iterar sobre elemetos iterables u objetos generator. Su sintaxis es mucho mas simple que en C/C++ o Java. En vez de definir el una variable de iteracion para aunmentar/disminuir, cada iteracion devuelve el objeto actual de la secuencia iterable.

**Nota:** La secuencia es recorrida en orden de aparicion.

Veamos unos ejemplos:

In [ ]:
for x in [3, 9, 12, 4]:
    print(x)

In [ ]:
prefix = "JKLMNOPQ"
suffix = "ack"
for letter in prefix:
    print(letter + suffix)

In [ ]:
for i in range(10):
    print(i**2)


`range` permite definir una lista de indices para recorrer en secuencia. Esto permite tambien definir una secuencia de numeros entre dos valores (no inclusivo) y secuencias con skips, similar a las notaciones de slicing:

**Note**: Similar a ``zip``, ``dict.keys`` y ``dict.values``, ``range`` retorna un objeto generador, el cual implica que slicing e iteracion no estan disponibles

In [ ]:
list(range(10))

`range`: Lista entre ``a`` y ``b - 1``:

In [ ]:
list(range(1, 20))

Con saltos de 3:

In [ ]:
list(range(7, 25, 3))


Notaciones de desempaquetado iterable puede ser usadas en un ciclo for, por ejemplo, ``enumerate`` funcion que nos permite recuperar el objeto iterado actual junto con su posicion:

In [ ]:
for i, x in enumerate([2, 3, 4]):
    print(i, x)


In [ ]:
for i, (a, b) in enumerate([(2, 3), (3, 4)]):
    print(i, a+b)

In [ ]:
for a, b in zip([3, 4], [5, 6]):
    print(a, b)

In [ ]:
duenios_mascotas = {
    'Laura'     : 'Max',
    'Juan'      : 'Kyra',
    'Catalina'  : 'Guadalupe',
    'Alejandro' : 'Thiago'
}
print('=> Los duenos de las mascotas son:')
for duenio in duenios_mascotas.keys():
    print(duenio)
    
print(40 * '-')
print('=> Las mascotas son:')
for mascota in duenios_mascotas.values():
    print(mascota)
    
print(40 * '-')
for duenio, mascota in duenios_mascotas.items():
    print(f'La mascota de {duenio} se llama {mascota}')
    

## Compension Listas/Diccionarios
Algunas veces, nosotros solo podemos usar un ciclo for para filtrar el contenido de una lista o crear un diccionario basado en otra lista. Sin embargo, Python nos brinda un manera mas facil para realizar este tipo de operaciones en una sola linea:

In [ ]:
a = []
for x in range(0, 20):
    if x % 2 == 0:
        a.append(x)
a

In [ ]:
a = [x for x in range(0, 20) if x % 2 == 0]  # Obtener todos los numeros pares
a

In [ ]:
a = []
for x in range(0, 20):
    if x % 2 == 0:
        a.append(x)
    else:
        a.append(0)
a

In [ ]:
a = [x if x % 2 == 0 else 0 for x in range(0, 20)] # En los indices impares poner 0
a

In [ ]:
a = [x if x % 2 == 0 else x**2 for x in range(0, 20)] # En los indices impares se eleva al cuadrado
a

Tambien podemos anidar esta notacion:

In [ ]:
a = []
for x in range(0, 5):
    temp_list = []
    for y in range(0, x):
        temp_list.append(0)
    a.append(temp_list)
a

In [ ]:
a = [[0 for y in range(0, x)] for x in range(0, 5)]
a

Podemos aplicar una notacion similar para crear diccionarios:

In [ ]:
d = {}
for i in range(0, 10):
    d[i] = i+10
d

In [ ]:
d = {i:i+10 for i in range(0, 10)}
d

## Problemas

### Problema 5

Construir un ciclo ``for`` que imprima todos los numeros pares de 1 a 100.

In [ ]:
# Escribir la solucion aqui


### Problema 6

Definir una funcion `es_primo` que tome un numero `x` y verifique si es divisible entre todos los numeros menores a `x`. Si lo es, entonces debe retornar `False` y si no `True`. Por ejemplo:

```python
es_primo(10)
False
```

```python
es_primo(17)
True
```

```python
es_primo(15)
False
```

```python
es_primo(23)
True
```

**Nota**:

Modificar la funcion `es_divisible_entre_n` para que en lugar de imprimir oraciones, retorne `True` o `False`, y despues utilizarla como parte del enunciado de la funcion `es_primo`.

In [ ]:
# Escribir la solucion aqui


### Problema 7

Optimizar la funcion anterior, respondiendo a la siguiente pregunta: Es necesario revisar todos los numeros menores a `x` para verificar si `x` es divisible entre todos ellos? Hasta que numero es en realidad necesario revisar?

Para ello, definir una nueva funcion `es_primo_veloz` y comparar los tiempos de ejecucion entre ella y `es_primo` usando el comando `%timeit`, asi:

```python
%timeit es_primo(600)
```

```ipython
%timeit es_primo_veloz(600)
```


In [ ]:
# Escribir la solucion aqui


### Problema 8

Definir una funcion `desv_est` que calcule la desviacion estandar de una lista, usando la formula:

$$s=\sqrt{\frac{\sum_{i=1}^{n}\left(x_{i}-\bar{x}\right)^{2}}{n-1}}$$

donde $\bar{x}$ es el promedio y $n$ es el numero total de datos.

Por ejemplo, la desviacion estandar de la siguiente lista:

In [ ]:
li = [48.38,  27.6 ,  32.46,  51.94,  47.43,  48.61,  34.38,  48.98,\
      48.86,  41.45,  56.55,  25.46,  27.03,  36.72,  48.03,  36.86,\
      42.58,  44.44,  56.12,  43.86,  44.42,  42.92,  41.43,  22.81,\
      36.55,  50.89,  29.93,  47.61,  63.91,  53.98,  42.64,  27.18,\
      29.93,  31.51]
def desv_est(lista):
    # YOUR CODE HERE
    raise NotImplementedError()


In [ ]:
assert(desv_est(li)==10.193054313544058)

debe dar como resultado

```python
desv_est(li)
10.193054313544058
```

# NumPy

Numpy es un paquete fundamental para la conputacion cientifica en Python. Contiene, entre muchas cosas:

- objeto de arreglos N-dimensionales
- funciones sofisticadas (de radiodifusion)
- herramientas de integracion C/C++ y herramientas de codigo Fortran 
- utiles capacidades de algebra lineal, transformada de Fourier, y numeros aleatorios

Ademas de los obvios usos cientificos, Numpy puede tambien ser usada de forma eficiente como contenedor multi-dimensional de datos genericos. Pueden definirse data-types de forma arbitraria. Esto permite a Numpy, sin problemas y rapidamente, integrarse con una amplia variedad de bases de datos.

In [ ]:
import numpy as np

In [ ]:
a = np.array([2,3,4]) #se genera un arreglo de numpy a partir de un arreglo de Python

In [ ]:
a

In [ ]:
a.dtype

In [ ]:
b = np.array([1.2, 3.5, 5.1])

In [ ]:
b.dtype

In [ ]:
a = np.array(1,2,3,4)

In [ ]:
a = np.array([1,2,3,4])
print(a)

In [ ]:
b = np.array([(1.5,2,3), (4,5,6)])
print(b)

In [ ]:
np.zeros((3, 4)) # se pueden generar arreglos de solo ceros

In [ ]:
c=np.ones((2, 3, 4), dtype=np.int16)  #se pueden generar arreglos de solo unos y se puede definir el tipo de dato de cualquier arreglo
print(c)
print(c.shape)

In [ ]:
np.arange(10, 30, 5) #se pueden generar arreglos de datos secuenciales, con el valor inicial, valor final y el paso.

In [ ]:
[i for i in range(10,30,5)] 

In [ ]:
np.arange(0, 2, 0.3) #se pueden usar pasos arbitrarios

In [ ]:
b = np.arange(12).reshape(4,3) #se puede cambiar las dimensiones de un arreglo simepre y cuando el tamaño concuerde con la cantidad de datos
print(b)

In [ ]:
A = np.array([[1,1], [0,1]])
print(A)
print()
B = np.array([[2,0], [3,4]])
print(B)
print()
print(A*B) #multiplicacion dato a dato
print()
print(A.dot(B)) #se realiza el producto punto de las dos matrices
print()
print(np.dot(A, B)) #se realiza el producto punto de las dos matrices

In [ ]:
np.vstack((A,B)) #concatenar dos arreglos de forma vertical (filas)

In [ ]:
b = np.random.random((2,3)) #genera una matriz con las ddimensiones especificadas con numeros aleatorios
print(b)

In [ ]:
b.sum(axis=0) # suma todos los datos de una matriz en un eje especificado

## Copias

#### No una copia

In [ ]:
a = np.arange(12)
b = a            # no se ha creado un nuevo objeto
print(a)
print(b)

In [ ]:
print(b is a)
b.shape = 3,4
a.shape

In [ ]:
print(a)
print(b)
print('---')
print(id(a))
print(id(b))

#### Copia profunda

In [ ]:
d = a.copy()
print(d is a)
print(d.base is a)

In [ ]:
d[0,0] = 9999
print(d)
print(a)

# Algebra lineal util

In [ ]:
X = np.array(
    [[1, 2, 3],
     [4, 5, 6],
     [7, 8, 9]]
)
eigvals, eigvecs = np.linalg.eig(X)
print('Eigenvalues: \n', eigvals)
print('Eigenvectors: \n', eigvecs)

In [ ]:
mat_pow = np.linalg.matrix_power(X, n=3)
print('x**3 = \n', mat_pow)

In [ ]:
gram_mat = np.matmul(X, X.T)
print('X * X^T = \n', gram_mat)

# Clases

Python es un muy buen lenguaje orientado a objetos (como Java), por lo que en el es muy sencillo definir y trabajar con clases. En las secciones anteriores hemos tenido la oportunidad de utilizar los metodos de cadenas y listas para resolver varios problemas, lo que indica que, en su implementacion, ambos tipos son clases, cosa que tambien es cierta para los demas (aun los numeros son clases!).

En esta seccion veremos como definir nuestras propias clases y vamos a aprender como asociarle sus correspondientes atributos y metodos.

## Definicion e inicializacion

En general, una clase es un objeto en el que se reunen varias funciones (llamadas metodos) y variables (llamadas atributos), con el objeto de que los metodos puedan compartir un mismo conjunto de datos, sobre el que puedan operar de cierta forma para llegar al resultado que desea el programador.

En Python las clases se definen con la palabra `class` y se inicializan usando el metodo ``__init__``, que es una funcion cuyo primer argumento **siempre** debe ser la palabra `self`. Los argumentos que vengan despues de `self` van a usarse para darle valores iniciales a los atributos de la clase.

Miremos como se hace con un ejemplo:

In [ ]:
class NumeroComplejo:
    def __init__(self, real, img):
        self.real = real
        self.img = img

Como se infiere de su nombre, esta clase se ha definido para representar numeros complejos, por lo que debe tener dos atributos: una parte real y una parte imaginaria. En este caso, estos estan dados por `real` e `img`, respectivamente.

----

**Nota**:

Es muy importante notar que para diferenciar los atributos de una clase de las variables locales del metodo, en Python todo atributo debe ir precedido de `self.`, como en `self.real` y `self.img`.

----

Ademas de `self`, podemos ver que `__init__` recibe los argumentos `real` e `img`, que se utilizan para inicializar los atributos mencionados.

Para crear una instancia de la clase es necesario llamarla por su nombre, con el numero de argumentos declarados en `__init__` (sin contar `self`) y asignarsela a una variable, asi:

In [ ]:
z = NumeroComplejo(1, 2)

Para comprobar que la inicializacion ha funcionado correctamente, podemos inspeccionar los atributos de la clase directamente:

In [ ]:
z.real

In [ ]:
z.img

De esta forma puede certificarse que, efectivamente, `z` es un numero complejo con parte real `1` y parte imaginaria `2`.

Una vez definida una instancia, tambien es posible modificar sus atributos por medio de asignacion, asi:

In [ ]:
z.real = 5

In [ ]:
z.real

## Metodos

Como ya dijimos, los metodos son funciones asociadas a una clase que operan sobre sus atributos. Por ejemplo, a la clase anterior le podemos añadir un metodo que calcule el modulo de un numero complejo con la formula:

$$\left| z \right|=\sqrt{\textrm{Re}\left(z\right)^{2}+\textrm{Im}\left(z\right)^{2}}$$

Para ello redefinimos `NumeroComplejo` para agregarle un nuevo metodo `modulo`, asi:

In [ ]:
class NumeroComplejo:
    def __init__(self, real, img):
        self.real = real
        self.img = img

    def modulo(self):
        return (self.real**2 + self.img**2)**(1/2)

**Nota**:

Al igual que para `__init__`, el primer argumento de todo metodo debe ser `self`, para indicar que hace parte de la clase.

Con esta nueva definicion obtenemos el siguiente resultado para el modulo del numero complejo definido a continuacion:

In [ ]:
z = NumeroComplejo(1, 2)

In [ ]:
z.modulo()

Aqui puede parecer un poco extraño que `modulo` se llame sin argumentos, cuando al definirlo en la clase se le habia pasado a `self` como primer argumento. Esto se debe a que `self` no es un argumento en si, sino que solo se usa para señalar que una funcion es un metodo de la clase, como ya se menciono.

Otra operacion que puede hacerse con numeros complejos es obtener su *conjugado*. El conjugado de un complejo $z$, es un nuevo numero complejo que se denota $\bar{z}$ y se define como

$$z=a+ib \longrightarrow \bar{z}=a-ib$$

Para obtener el conjugado podemos entonces agregar un nuevo metodo a nuestra clase, de la siguiente forma:

In [ ]:
class NumeroComplejo(object):
    def __init__(self, real, img):
        self.real = real
        self.img = img

    def modulo(self):
        return (self.real**2 + self.img**2)**(1/2)

    def conjugado(self):
        return NumeroComplejo(self.real, -self.img)

Para calcular el conjugado de un numero complejo `z`, solo debemos llamar el metodo correspondiente y asignaserlo a una nueva variable `z1` (en caso de que deseemos usarlo mas tarde):

In [ ]:
z = NumeroComplejo(1, 2)

In [ ]:
z1 = z.conjugado()

In [ ]:
z1.real

In [ ]:
z1.img

Finalmente, vamos a añadir un metodo que retorne el producto de dos numeros complejos. Dados dos numeros

$$
z = a + ib \\
w = c + id
$$

su producto esta dado por:

$$z \times w = (ac - bd) + i(ad + bc)$$

Para ello podemos escribir el siguiente metodo, llamado `producto`, en nuestra clase:

In [ ]:
class NumeroComplejo(object):
    def __init__(self, real, img):
        self.real = real
        self.img = img

    def modulo(self):
        return (self.real**2 + self.img**2)**(1/2)

    def conjugado(self):
        return NumeroComplejo(self.real, -self.img)

    def producto(self, w):
        real = self.real * w.real - self.img * w.img
        img = self.real * w.img + self.img * w.real
        return NumeroComplejo(real, img)

In [ ]:
z = NumeroComplejo(1, 2)
w = NumeroComplejo(4, -7)

In [ ]:
x = z.producto(w)

In [ ]:
x.real

In [ ]:
x.img

Para comprobar que `producto` esta funcionando correctamente podemos usar la siguiente formula, que relaciona el modulo de un numero complejo con su conjugado:

$$\left| z \right| = \sqrt{\textrm{Re} \left( z \times \bar{z} \right)}$$

In [ ]:
z2 = z.producto(z.conjugado())

In [ ]:
(z2.real)**(1/2) == z.modulo()

La especificacion de clases de Python permite sobrecargar otras funciones comunes y operadores.

**Para obtener mayor informacion, visitar:** https://docs.python.org/3/library/operator.html y https://docs.python.org/3/reference/datamodel.html

## Libreria Python Imaging

La libreria de Python Imaging, o PIL abreviado, es una de las librerias central de la manipulacion de imagenes en Python. Desafortunadamente, su desarrollo se ha estancado, su ultima atualizacion fue en 2009.

Por suerte, hay una forma que PIL que es desarrollada activamente llamada Pillow, es mas facil de installar y corre en todos los sistemas operativos, y es compatible con Python 3

## Instalacion

Si usted no tiene PIL instalado en su maquina, simplemente escriba el siguiente comando en la terminal, y estara todo hecho:

`pip install pillow` o si este no funciona, `pip install --user pillow`

In [ ]:
from PIL import Image

In [ ]:
import requests
from io import BytesIO
url='https://dab1nmslvvntp.cloudfront.net/wp-content/uploads/2015/09/1441780452brick-house.png'
response = requests.get(url)
img = Image.open(BytesIO(response.content))

In [ ]:
img.show()

In [ ]:
img

In [ ]:
img.save('imgs/brick-house.png')

----
### El archivo de brick-house.png es una imagen a color. Para convertirlo a una imagen en escala de grises, carguela y luego guarde la nueva imagen en escala de grises, puede seguir el siguiente codigo:

In [ ]:
from PIL import Image
img_gs = img.convert('L')
img_gs
img_gs.save('imgs/brick-house-gs.png')

Mas informacion en http://pillow.readthedocs.io/en/3.1.x/reference/Image.html

## Resize

In [ ]:
new_img = img.resize((256,256))
new_img

In [ ]:
import numpy as np
np.array(new_img)

In [ ]:
np.array(new_img).shape

### Bibliografia
- https://www.sitepoint.com/manipulating-images-with-the-python-imaging-library/

# Matplotlib

Matplotlib es una libreria de graficas 2D que produce una figura de calidad en diferentes formatos y ambientes interactivos en todas las plataformas. Matplotlib puede ser usado en codigos de Python, celdas de IPython, Jupyter notebook, servidores de aplicaciones web, y como herramientas de interfaces graficas de usuario.

Matplotlib intenta hacer faciles las cosas faciles y posibles las cosas dificiles. Usted puede generar graficas, histogramas, espectros de potencia, graficas de barra, graficas de error, graficos de dispersion, etc. con unas pocas lineas de codigo. Observe unos ejemplos de graficos y galeria de miniaturas.

Para realizar una grafica simple, el modulo pyplot provee una interfaz similar a la de MATLAB, en particular cuando se convina con IPython. Para los usuarios experimentados, se tiene control total de estilos de linea, propiedades de fuente, propiedades de ejes, etc. por medio de interfaces orientadas a objetos o un conjunto de funciones similares a los usuarios de MATLAB.

## Instalacion

Si no tiene matplotlib installada, corra el siguiente comando en la terminal y todo estara listo:

`pip install matplotlib` o si este no funciona, `pip install --user matplotlib`

Ahora, vamos a abrir una imagen, tambien con esta librería, para poder usarla a lo largo de esta explicación con las siguientes lineas:

In [ ]:
import matplotlib.pyplot as plt

url = 'imgs/brick-house.png'
img = plt.imread(url) #PNG file!

Ahora que tenemos la imagen podemos usar la función ``imshow()`` para poder graficar la imagen.

In [ ]:
imgplot = plt.imshow(img)

Además de eso, como las imagenes son arreglos con 3 canales diferentes podemos graficar un unico canal de estos extrayendolo y luego usando la misma funcion que usamos anteriormente

In [ ]:
lum_img = img[:,:,0]
plt.imshow(lum_img)

Ahora, como se dieron cuenta por defecto viene un mapa de color que va desde el azul hasta el amarillo. Pero generalmente este mapa de color no es el mapa que queremos usar, sino uno que haga ver la imagen en escala de grises. Para esto, vamos a graficar nuevamente ``lum_img`` con un parametro extra: ``cmap``

In [ ]:
plt.imshow(lum_img,cmap='gray')

## Subplots
Uno de las herramientas que mas se usan y mas usarán es la realización de subplots. Existen muchas funciones diferentes que pueden usar para poder tener un subplot de la forma que ustedes prefieran, aqui podrán ver las mas basicas.

En primer lugar es importante saber como declarar un subplot, la forma mas usada generlamente es ``fig, ax = plt.subplots()`` donde los dos primeros parametros son el numero de filas y numero de columnas. Este trae consigo una ventaja muy grande y es que nos permite llenarlo al interior de un for y manejarlo como si se tratara de un arreglo. A continuación podemos ver un ejemplo donde graficaremos la misma imagen en un subplot de 3x4 dentro de un for

In [ ]:
fig, ax = plt.subplots(3,4)
for i in range(3):
    for j in range(4):
        ax[i,j].imshow(lum_img,cmap='gray')

Sin embargo, como pueden notar, tiene todavía los ejes, algo que no queremos, por lo cual podemos usar la funcion ``ax[].axis('off')`` para apagarlos

In [ ]:
fig, ax = plt.subplots(3,4)
for i in range(3):
    for j in range(4):
        ax[i,j].imshow(lum_img,cmap='gray')
        ax[i,j].axis('off')

La ventaja que tiene esta forma es justamente que nos permite llenarlo dentro de un recorrido, asi que no nos tocaría poner 81 lineas para graficar un subplot de 9x9, sino que podemos hacerlo en 5 lineas

In [ ]:
plt.rcParams['figure.figsize'] = [10, 15] # este es un parametro que nos sirve para escalar el tamaño de las imagenes
fig, ax = plt.subplots(9,9)
for i in range(9):
    for j in range(9):
        ax[i,j].imshow(lum_img,cmap='gray')
        ax[i,j].axis('off')

Finalmente, lo unico que puede que le falte a nuestro subplot son titulos, ya que muchas veces tienen información crucial sobre la imagen que no podemos ignorar. Para esto usaremos el comando ``ax[].set_title()``

In [ ]:
fig, ax = plt.subplots(9,9)
for i in range(9):
    for j in range(9):
        ax[i,j].imshow(lum_img,cmap='gray')
        ax[i,j].axis('off')
        ax[i,j].set_title(f'casa {i*9+j+1}')

### Bibliografia
- https://matplotlib.org/devdocs/index.html
- https://matplotlib.org/1.3.1/users/image_tutorial.html

# OpenSource Computer Vision

OpenSource Computer Vision, mas conocida como OpenCV, es un software de manipulacion y procesamiento de imagenes mas avanzada que PIL. Esta implementado en muchos lenguajes y es apliamente usada.

## Instalacion

En Python, el procesamiento de imagenes usando OpenCV es implementada usando los modulos de cv2 y Numpy. Hay dos opciones para instalar opencv, la facil y la no tan facil. 

Facil: `pip install opencv-python` 

Menos facil: Las [instrucciones de instalacion de OpenCV](http://docs.opencv.org/2.4/doc/tutorials/introduction/table_of_content_introduction/table_of_content_introduction.html#table-of-content-introduction) podria guiarte a traves de la configuracion del proyecto usted mismo.

In [ ]:
import matplotlib.pyplot as plt
import cv2
import numpy as np

In [ ]:
# Leer la imagen
# Escribir el path hacia la imagen 'beagle.jpg' de acuerdo a la distribucion de archivos.
img = cv2.imread('.../Python-tutorial/imgs/beagle.jpg')

# Hay un metodo de hacerlo mas facil y de forma global?

In [ ]:
plt.imshow(img)

In [ ]:
plt.imshow(img[:,:,(2,1,0)])

In [ ]:
#Mostrar la imagen
cv2.imshow('image',img)
cv2.waitKey(0)
cv2.destroyAllWindows()

# NO CIERRE LA VENTANA DE WINDOWS USANDO LA 'X' SOLO PRECIONE CUALQUIER TECLA!

In [ ]:
#Aplicar el filtro de escala de grises
gray = cv2.cvtColor(img, cv2.COLOR_BGR2GRAY)
gray = cv2.resize(gray, (224,224))
cv2.imshow('image',gray)
cv2.waitKey(0)
cv2.destroyAllWindows()

#Guardar la imagen en escala de grises
#cv2.imwrite('graytest.jpg',gray)

# Pickle

Pickle es un modulo de Python que usa protocolos binarios para empauqetar y desempaquetar objetos. Esta herramienta nos ayudara a guardar matrices, arreglos, objetos, modelos, etc.

In [ ]:
import pickle
import numpy as np

* **Nota**: Los archivos se pueden llamar como se desee, pueden agregarse las extensiones como .pkl o .mat pero no influye en nada.

In [ ]:
a=np.random.random((3,3))
with open("random_matrix.pkl","wb") as f: 
    pickle.dump(a,f) #para escribir elementos se usa "wb" y pickle.dump
a

* **Nota**: Los archivos que fueron guardados con pickle solo pueden abrirse con pickle, sin importar que sean .mat, .png o cualquier otra extencion.

In [ ]:
with open("random_matrix.pkl", "rb") as f: 
    b=pickle.load(f) #para leer archivos se usa "rb" y pickle.load, se debe guardar esta en una variable para su uso
b

# Paths


En el manejo de archivos se tienen varias formas de acceder a un docuemento dentro de una carpeta, se pueden usar paths absolutos o paths relativos. Los paths absolutos no se recomiendan ya que son especificos para el computador del usuario, por lo que volvera inutil el codigo en el computador de alguien mas. Los paths relativos se pueden contruir a partir de una estructura asumida de los archivos. 

La forma mas facil de acceder a un archivo es usando su ruta, relativa al script que se esté usando, sin importar el formato o herramienta que se use. Asi es como hemos accedido a todos los archivos que hemos usado en este tutorial. Es decir ``imgs/brick-house.png``, lo cual se puede leer como que dentro de la carpeta ``imgs`` existe un archivo de tipo ``png`` llamado ``brick-house``. Sin embargo, pueden suceder inconvenientes al usar este tipo de notación por lo que no todos los sistemas operativos usan este tipo de separador para las rutas. Algunos usan ``\\``,``\`` o ``//`` como separador, y tenemos que estar preparados para que nuestro codigo sea facil de usar para todos los usuarios sin importar su sistema operativo.

## Os.path.join

El metodo ``path.join`` de la libreria ``os`` nos permite crear paths de los archivos y acceder a estos de forma sencilla, usando el separador usado por el sistema operativo del computador que se esté usando. Veamos un ejemplo con la ruta anterior:

In [ ]:
import os
from matplotlib import pyplot as plt

ruta = os.path.join("imgs","brick-house.png")

Y como podemos ver funciona igual que con la ruta anterior.

In [ ]:
plt.imshow(plt.imread(ruta))

## Bases de datos
Esto funciona bien para una imagen, sin embargo, cuando vayamos a usar bases de datos de miles de imagenes no vamos a poder hacer linea por linea las rutas a todas las imagenes de la base de datos y muchas veces no podemos abrirlas por medio de un for ya que no siempre los nombres cumplen formatos. Para este proposito existe una librería llamada glob la cual nos permite justamente recopilar todos los **archivos o carpetas** que cumplan un formato que nosotros le especifiquemos. Por ejemplo podemos conocer todo lo que hay en la carpeta en la que estamos parados gracias a esta como en el ejemplo a continuación.

**Datos importantes:** Los asteriscos ``*`` cuentan como cualquier nombre de archivo o carpeta. Usar dos puntos ``..`` significa que se da un paso atras en la ruta(util para acceder cosas que estén afuera de la carpeta en la que se está)

In [ ]:
from glob import glob

print(glob('*'))

Por ejemplo, digamos que tenemos una base de datos como la que tenemos en la carpeta database, que son imagenes donde la mitad tienen el nombre negro y la otra mitad blanco con un numero, ademas están inmersas en carpetas de entrenamiento, validación y prueba. Un primer ejercicio puede ser revisar cuantas hay en total en la carpeta de train. Ya que sabemos en que formato estan (``jpg``) podemos acceder a ellas usando ``glob`` en combinacion de ``os.path.join`` con el siguiente comando:

In [ ]:
print(glob(os.path.join('database','train','*.jpg')))
print(f"largo: {len(glob(os.path.join('database','train','*.jpg')))}")

Como pudieron ver se puede hacer de forma relativamente simple, usando el asterisco como un string sin formato especificado. Ademas podemos hacerlo con las carpetas, en caso que querramos obtener las rutas de todas las imagenes de la base de datos sin importar su carpeta, usando un asterisco en el lugar donde va la carpeta

In [ ]:
print(glob(os.path.join('database','*','*.jpg')))
print(f"largo: {len(glob(os.path.join('database','*','*.jpg')))}")

Ahora bien, si por ejemplo quisieramos obtener unicamente las imagenes que están anotadas como negras, podemos tambien usar el asterisco acompañado de otro texto para poder tener las rutas que cumplan con un formato que deseemos

In [ ]:
print(glob(os.path.join('database','*','Negra-*.jpg')))
print(f"largo: {len(glob(os.path.join('database','*','Negra-*.jpg')))}")

¡Inclusive podemos ser mas especificos y buscar las imagenes cuyo numero perteneza a los 90, independiente de nombre o carpeta!

In [ ]:
print(glob(os.path.join('database','*','*-9*.jpg')))
print(f"largo: {len(glob(os.path.join('database','*','*-9*.jpg')))}")

Estas funciones que acabamos de apreder, acompañados de un buen manejo de strings puden ayudarnos a obtener cualquier ruta que queramos de forma automatica siempre que se cumpla con un formato especifico. Una de las funciones mas utiles es la funcion ``split()`` nativa de los strings, sin embargo, si vamos a separar rutas es necesario saber el separador. Para esto podemos usar ``os.path.sep`` para obtener el separador especifico. Veamos un ejemplo.

In [ ]:
ruta = os.path.join("imgs","brick-house.png")
print(f'la ruta es: {ruta}')
print(f'el separador es: {os.path.sep}')
print(f'la ruta separada es: {ruta.split(os.path.sep)}')


# Skimage

Skimage es un modulo que cuenta con una coleccion de herramientas de procesamiento de imagenes. Puede contar con herramientas simialres a opencv y suele ser mas intuitiva de usar.

# Instalacion

Para su facil instalacion corra en la terminal el siguiente comando:

``pip install scikit-image``

In [ ]:
import skimage
from skimage import io
import os
from matplotlib import pyplot as plt

In [ ]:
im=io.imread(os.path.join("imgs","brick-house.png")) #la lectura de las imagenes
plt.imshow(im)

Hay multiples funciones de ecualizacion, deteccion de maximos, operaciones morfologicas y segmentacion que usaremos mas adelante. Hechenle un ojo a la docuemntacion de los metodos!